modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/SqlRowHandler.java

```diff
@@ -0,0 +1,260 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.exec;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import org.apache.ignite.internal.schema.BinaryTuple;
+import org.apache.ignite.internal.schema.NativeType;
+import org.apache.ignite.internal.schema.row.InternalTuple;
+import org.apache.ignite.internal.sql.engine.exec.SqlRowHandler.RowWrapper;
+import org.apache.ignite.internal.sql.engine.exec.row.RowSchema;
+import org.apache.ignite.internal.sql.engine.exec.row.RowSchema.Builder;
+import org.apache.ignite.internal.sql.engine.exec.row.RowSchemaTypes;
+import org.apache.ignite.internal.sql.engine.exec.row.TypeSpec;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.sql.engine.util.TypeUtils;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Handler that uses a {@link RowWrapper wrapper} to operate with two types of row implementations.
+ *
+ * <ul>
+ *     <li>{@link ObjectsArrayRowWrapper Objects array}</li>
+ *     <li>{@link BinaryTupleRowWrapper Binary tuple}</li>
+ * </ul>
+ *
+ * <p>Each kind of rows is serialized to the same binary tuple format
+ * using the {@link #toByteBuffer(RowWrapper) toByteBuffer} method.
+ *
+ * <p>Factory methods {@link RowFactory#wrap(InternalTuple) wrap(InternalTuple)} and
+ * {@link RowFactory#create(ByteBuffer) create(ByteBuffer)} allow create rows without
+ * any additional conversions. But the fields in binary tuple must match the
+ * factory {@link RowSchema row schema}.
+ */
+public class SqlRowHandler implements RowHandler<RowWrapper> {
+    public static final RowHandler<RowWrapper> INSTANCE = new SqlRowHandler();
+
+    private SqlRowHandler() {
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @Nullable Object get(int field, RowWrapper row) {
+        return row.get(field);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void set(int field, RowWrapper row, @Nullable Object val) {
+        row.set(field, val);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public RowWrapper concat(RowWrapper left, RowWrapper right) {
+        int leftLen = left.columnsCount();
+        int rightLen = right.columnsCount();
+        List<TypeSpec> leftTypes = left.rowSchema().fields();
+        List<TypeSpec> rightTypes = right.rowSchema().fields();
+
+        Object[] values = new Object[leftLen + rightLen];
+        Builder schemaBuilder = RowSchema.builder();
+
+        for (int i = 0; i < leftLen; i++) {
+            values[i] = left.get(i);
+            schemaBuilder.addField(leftTypes.get(i));
+        }
+
+        for (int i = 0; i < rightLen; i++) {
+            values[leftLen + i] = right.get(i);
```

Comment:
no need to call additional offset here : leftLen + i
you just can initialize global int i = 0; and increment it in further for loops.

Reply:
We need two indices anyway.  I don't think it worth to replace `leftLen + i` to something like 'index++' and declare `index` out of cycle.